cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,681 @@
+                                             edgelist_dsts.begin() + edgelist_srcs.size()),
+                   get_dataframe_buffer_begin(invalid_edges_buffer));
+
+      // resize the 'edgelist_srcs' and 'edgelist_dst'
+      edgelist_srcs.resize(num_valid_edges, handle.get_stream());
+      edgelist_dsts.resize(num_valid_edges, handle.get_stream());
+      num_triangles.resize(num_valid_edges, handle.get_stream());
+
+      // case 1: unroll (q, r)
+      // For each (q, r) edges to unroll, find the incoming edges to 'r' let's say from 'p' and
+      // create the pair (p, q)
+      cugraph::find_unroll_p_r_and_q_r_edges<vertex_t,
+                                             edge_t,
+                                             decltype(transposed_edge_first),
+                                             decltype(edge_triangle_count_pair_first),
+                                             false>(handle,
+                                                    cur_graph_view,
+                                                    0,  // 0 for (q, r) and 1 for (p, r) edges
+                                                    num_invalid_edges,
+                                                    std::move(invalid_edges_buffer),
+                                                    std::move(edgelist_srcs),
+                                                    std::move(edgelist_dsts),
+                                                    std::move(num_triangles),
+                                                    transposed_edge_first,
+                                                    edge_triangle_count_pair_first);
+
+      // case 2: unroll (p, r)
+      cugraph::find_unroll_p_r_and_q_r_edges<vertex_t,
+                                             edge_t,
+                                             decltype(transposed_edge_first),
+                                             decltype(edge_triangle_count_pair_first),
+                                             false>(handle,
+                                                    cur_graph_view,
+                                                    1,  // 0 for (q, r) and 1 for (p, r) edges
+                                                    num_invalid_edges,
+                                                    std::move(invalid_edges_buffer),
+                                                    std::move(edgelist_srcs),
+                                                    std::move(edgelist_dsts),
+                                                    std::move(num_triangles),
+                                                    transposed_edge_first,
+                                                    edge_triangle_count_pair_first);
+
+      cugraph::fill_edge_property(handle, cur_graph_view, false, edge_mask);
+      edges_with_triangles.clear();  // FIXME: is this needed?
+      edges_with_triangles.insert(
+        edgelist_srcs.begin(), edgelist_srcs.begin(), edgelist_dsts.end());
```

Comment:
And you can't mask out all the invalid (but with > 0 triangle count) edges. You can only mask out the edges with triangle count 0.

See Algorithm 1 of https://www.osti.gov/servlets/purl/1481088

Reply:
I think we should perform Case 1, 2, 3 first, then remove (mask-out) 0 triangle count edges at the end.